iroh/src/downloader/get.rs

```diff
@@ -0,0 +1,561 @@
+            Read(e) => e.into(),
+            e @ CustomRequestTooBig => {
+                // something wrong with the request itself
+                FailureAction::AbortRequest(e.into())
+            }
+            e @ Eof => {
+                // TODO(@divma): unsure about this based on docs
+                FailureAction::RetryLater(e.into())
+            }
+            e @ PostcardDe(_) => {
+                // serialization errors can't be recovered
+                FailureAction::AbortRequest(e.into())
+            }
+            e @ Io(_) => {
+                // io errors are likely recoverable
+                FailureAction::RetryLater(e.into())
+            }
+        }
+    }
+}
+
+impl From<iroh_bytes::get::fsm::AtBlobHeaderNextError> for FailureAction {
+    fn from(value: iroh_bytes::get::fsm::AtBlobHeaderNextError) -> Self {
+        use iroh_bytes::get::fsm::AtBlobHeaderNextError::*;
+        match value {
+            e @ NotFound => {
+                // > This indicates that the provider does not have the requested data.
+                // peer might have the data later, simply retry it
+                FailureAction::RetryLater(e.into())
+            }
+            e @ InvalidQueryRange => {
+                // we are doing something wrong with this request, drop it
+                FailureAction::AbortRequest(e.into())
+            }
+            Read(e) => e.into(),
+            e @ Io(_) => {
+                // io errors are likely recoverable
+                FailureAction::RetryLater(e.into())
+            }
+        }
+    }
+}
+
+impl From<iroh_bytes::get::fsm::DecodeError> for FailureAction {
+    fn from(value: iroh_bytes::get::fsm::DecodeError) -> Self {
+        use get::fsm::DecodeError::*;
+
+        match value {
+            e @ NotFound => FailureAction::RetryLater(e.into()),
+            e @ ParentNotFound(_) => FailureAction::RetryLater(e.into()),
+            e @ LeafNotFound(_) => FailureAction::RetryLater(e.into()),
+            e @ ParentHashMismatch(_) => {
+                // TODO(@divma): did the peer sent wrong data? is it corrupted? did we sent a wrong
+                // request?
+                FailureAction::AbortRequest(e.into())
+            }
+            e @ LeafHashMismatch(_) => {
+                // TODO(@divma): did the peer sent wrong data? is it corrupted? did we sent a wrong
+                // request?
+                FailureAction::AbortRequest(e.into())
+            }
+            e @ InvalidQueryRange => FailureAction::AbortRequest(e.into()),
+            Read(e) => e.into(),
+            Io(e) => e.into(),
+        }
+    }
+}
+
+impl From<std::io::Error> for FailureAction {
+    fn from(value: std::io::Error) -> Self {
+        // generally consider io errors recoverable
+        // we might want to revisit this at some point
+        FailureAction::RetryLater(value.into())
+    }
+}
+
+/// Get a blob or collection
+pub async fn get<D: Store, C: CollectionParser>(
```

Comment:
@rklaehn would be good if you could review the below `get_*` functions, given you are the most familiar with fetching blob